### PR TITLE
feat: autonomous SSM Bastion

### PR DESCRIPTION
**Pull request by DjinnS**

### CODE REVIEW

File: examples/custom_ssh_key/main.tf [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+module "my_ssm_bastion" {
+  source = "../.."
+
+  ssm_logging_bucket_name = aws_s3_bucket.ssm_logs.id
+  security_groups         = [aws_security_group.bastion_ssm.id]
+  vpc_zone_identifier     = module.my_vpc.private_subnets_ids
+
+  manage_ssm_user_ssh_key = false
+  custom_ssm_user_public_key = tls_private_key.ssm_user.public_key_openssh
```
Comment (Matthieulvt) [inline examples/custom_ssh_key/main.tf:45]: l'alignement des = n'est pas bon sur la ligne 44.

File: launch_template.tf [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+# shared for all ops while isn't possible to open remote tunnels from AWS SSM
+# THIS WILL BE REMOVE WHEN SSM REMOTE TUNNEL WILL BE IMPLEMENTED
+resource "tls_private_key" "ssm_user" {
+  count = var.manage_ssm_user_ssh_key ? 1 : 0
+
+  algorithm = "RSA"
+  rsa_bits  = "4096"
+}
+
+# an SSH key for ec2-user
```
Comment (Matthieulvt) [inline launch_template.tf:29]: l30 -> l41 Should this be removed ? (either uncommented or deleted)

File: main.tf [hunk 3]
```diff
@@ -1,6 +1,35 @@
-# Here you can reference 2 type of terraform objects :
-# 1. Ressources from you provider of choice
-# 2. Modules from official repositories which include modules from the following github organizations
-#     - AWS: https://github.com/terraform-aws-modules
-#     - GCP: https://github.com/terraform-google-modules
-#     - Azure: https://github.com/Azure
+
```
Comment (Matthieulvt) [inline main.tf]: delete this line

File: iam.tf [hunk 4]
```diff
@@ -0,0 +1,76 @@
+# default IAM for our Bastion
+# merged with config provided to init the module
+
```
Comment (Matthieulvt) [inline iam.tf]: delete this line and maybe description above ?

File: launch_template.tf [hunk 5]
```diff
@@ -0,0 +1,85 @@
+
```
Comment (Matthieulvt) [inline launch_template.tf]: delete this line

File: ssm_logging.tf [hunk 6]
```diff
@@ -0,0 +1,33 @@
+# enable logging to s3 for session manager
+
+# if you encounterd this error:
+#
+# Error: Error creating SSM document: DocumentAlreadyExists: Document with same name SSM-SessionManagerRunShell already exists
+#
+# delete the document first:
+# $ aws --profile $AWS_PROFILE ssm delete-document --name  SSM-SessionManagerRunShell
+#
+# Source: https://github.com/hashicorp/terraform-provider-aws/issues/7516
```
Comment (Matthieulvt) [inline ssm_logging.tf:10]: l1 -> l10 : This should be in the README #Troubleshooting

File: variables.tf [hunk 7]
```diff
@@ -1,20 +1,174 @@
-variable "example_of_required_variable" {
+#
```
Comment (Matthieulvt) [inline variables.tf:1]: delete this line

File: variables.tf [hunk 8]
```diff
@@ -1,20 +1,174 @@
...
+  default     = true
+}
+
+variable "custom_ssm_user_public_key" {
+  description = "The public key to use for the ssm-user user"
+  type        = string
+  default     = ""
+}
+
+#variable "manage_ec2_user_ssh_key" {
```
Comment (Matthieulvt) [inline variables.tf:161]: l 159 ->168 : Uncomment or remove

File: variables.tf [hunk 9]
```diff
@@ -1,20 +1,174 @@
...
+#
+variable "ami_id" {
+  description = "The AMI from which to launch the instance"
   type        = string
-  description = "Short description of the variable"
-  default     = "default_value"
+  default     = ""
+}
+
+#variable "key_name" {
```
Comment (Matthieulvt) [inline variables.tf:33]: l 34-> l38 : Uncomment or remove

File: variables.tf [hunk 10]
```diff
@@ -1,20 +1,174 @@
...
+  type        = list(any)
+  default = [
+    "GroupMinSize", "GroupMaxSize", "GroupDesiredCapacity",
+    "GroupInServiceInstances", "GroupPendingInstances", "GroupStandbyInstances",
+    "GroupTerminatingInstances", "GroupTotalInstances"
+  ]
+}
+
+variable "manage_ssm_user_ssh_key" {
+  description = "Wether you want to let the module manage the ssh key for the ssm-user, if set to false you need to set `custom_ssm_user_public_key` "
```
Comment (Matthieulvt) [inline variables.tf]: Whether*

File: variables.tf [hunk 11]
```diff
@@ -1,20 +1,174 @@
...
+#  type        = string
+#  default     = ""
+#}
 
-  validation {
-    condition     = length(var.example_with_validation) >= 2
-    error_message = "Error message which explains what's required and finished with a dot ."
-  }
+variable "add_ssm_user_to_sudoers" {
+  description = "Wether you want to add the ssm_user to sudoers"
```
Comment (Matthieulvt) [inline variables.tf]: *Whether

File: variables.tf [hunk 12]
```diff
@@ -1,20 +1,174 @@
...
+#
+variable "ssm_logging_bucket_name" {
+  description = "SSM Logging Bucket name"
   type        = string
-  description = "Short description of the variable"
 }
 
-variable "example_of_variable_with_default_value" {
+variable "ssm_logging_bucket_encryption" {
+  description = "Wether the Amazon S3 bucket you specified in the s3BucketName input must be encrypted"
```
Comment (Matthieulvt) [inline variables.tf]: Whether

File: variables.tf [hunk 13]
```diff
@@ -1,20 +1,174 @@
...
+}
+
+variable "custom_ssm_user_public_key" {
+  description = "The public key to use for the ssm-user user"
+  type        = string
+  default     = ""
+}
+
+#variable "manage_ec2_user_ssh_key" {
+#  description = "Wether you want to let the module manage the ssh key for the ec2-user, if set to false you need to set `custom_ec2_user_public_key` "
```
Comment (Matthieulvt) [inline variables.tf]: Whether

File: launch_template.tf (hunk 2 above)
Comment (DjinnS) [reply launch_template.tf:29]: I want to keep this line for the future (when the new release of ssm-agent will support remote forward)

File: iam.tf (hunk 4 above)
Comment (DjinnS) [reply iam.tf]: Why ? It's the description of the purpose of the file

File: variables.tf (hunk 9 above)
Comment (DjinnS) [reply variables.tf:33]: I want to keep this line for the future (when the new release of ssm-agent will support remote forward)

**Pull request merged**